### PR TITLE
Include required properties in `--default-values-file-output`

### PR DESCRIPTION
**Pull request by praveenrewar**
#### What this PR does / why we need it:
Include required properties in --default-values-file-output

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
`--default-values-file-output` will also include required properties.
```

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: cli/test/e2e/env.go [hunk 1]
```diff
@@ -44,6 +44,14 @@ func (e Env) Validate(t *testing.T) {
 		errStrs = append(errStrs, "Expected Namespace to be non-empty (hint: kubectl create namespace kctrl-test; export KCTRL_E2E_NAMESPACE=kctrl-test)")
 	}
 
+	if len(errStrs) > 0 {
+		t.Fatalf("%s", strings.Join(errStrs, "\n"))
+	}
+}
+
+func (e Env) ValidateImageURL(t *testing.T) {
+	errStrs := []string{}
```
Comment (100mik) [inline cli/test/e2e/env.go]: I think we can do without errStrs here because it won't be checked with the other parameters anyway.

File: cli/pkg/kctrl/cmd/package/available/default_values.go [hunk 2]
```diff
@@ -138,28 +137,36 @@ func (s PackageSchema) isKindInt(src interface{}) bool {
...
+	if structural.Type != "object" {
+		return append(b, true)
+	}
 	for _, v := range structural.Properties {
-		// return true if there is a non-nested(not object) with a default value
-		if v.Type != "object" && v.Default.Object != nil {
-			b = append(b, true)
-			return b
+		// return true if there is a non-nested(not object)
+		if v.Type != "object" {
```
Comment (100mik) [inline cli/pkg/kctrl/cmd/package/available/default_values.go]: Are we not able to check if the non-object field is required at this point?

File: cli/test/e2e/release_with_changed_spec_test.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ func TestPackageReleaseWithChangedSpec(t *testing.T) {
 	env := BuildEnv(t)
 	logger := Logger{}
 	kappCtrl := Kctrl{t, env.Namespace, env.KctrlBinaryPath, logger}
+	env.ValidateImageURL(t)
```
Comment (100mik) [inline cli/test/e2e/release_with_changed_spec_test.go]: not sure about running validations on image env like this, mainly because the user will only hit this when this tests is run. Should we move this to a separate PR? Setting an image that is not used does not seem that much of an overhead 🤔

File: cli/test/e2e/env.go (hunk 1 above)
Comment (praveenrewar) [reply cli/test/e2e/env.go]: Reverted this change.

File: cli/test/e2e/release_with_changed_spec_test.go (hunk 3 above)
Comment (praveenrewar) [reply cli/test/e2e/release_with_changed_spec_test.go]: Reverted this change.

File: cli/pkg/kctrl/cmd/package/available/default_values.go (hunk 2 above)
Comment (praveenrewar) [reply cli/pkg/kctrl/cmd/package/available/default_values.go]: There's no required field on the properties. The required field information is available separately on the values schema.

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/available/default_values.go]: The OpenAPI schema does reflect which fields are required right? 🤔 
(Or is it always false)

Comment (praveenrewar) [reply cli/pkg/kctrl/cmd/package/available/default_values.go]: This is how it's defined:
```
valuesSchema:
    openAPIv3:
      properties:
        testField:
          description: The test field
          type: string
        requiredField1:
          description: The required field
          type: string
        requiredField2:
          description: Another required field
          type: string
      required:
      - requiredField1
      - requiredField2
```

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/available/default_values.go]: Correct, so should we not emit only the ones that are in the required list? 
(I thought that was the intention)

Comment (praveenrewar) [reply cli/pkg/kctrl/cmd/package/available/default_values.go]: That wasn't the intention of the PR, but do you think we should take that path?
I felt that it would further complicate the code.

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/available/default_values.go]: I wonder how it affects larger packages with a decent number of fields that are generally unused

File: cli/hack/test-e2e.sh [hunk 4]
```diff
@@ -1,11 +1,18 @@
 #!/bin/bash
 
-set -e -x -u
+set -e -x -o pipefail
```
Comment (100mik) [inline cli/hack/test-e2e.sh:3]: Just to be sure, we are removing `-u` because we do not want to error out if ns env variable is not set correct?

File: cli/pkg/kctrl/cmd/package/available/default_values.go [hunk 5]
```diff
@@ -32,6 +36,10 @@ func (s PackageSchema) DefaultValues() ([]byte, error) {
 		return nil, err
 	}
 
+	if len(ss.Properties) == 0 {
```
Comment (100mik) [inline cli/pkg/kctrl/cmd/package/available/default_values.go:39]: nit: refactor: `ss` => `structuralSchema`

**Review by 100mik — COMMENTED** [review]
Some thoughts.

Since this PR has a few refactors/improvements separated by commits I was wondering if it would be worth cleaning up some of the nomenclature in `default_values.go`

The nit mentioned aside I think we should rename the slice of booleans (`b`) and even `x` (this seems to have different meanings in different scopes?)
`k` = `key` is the conclusion I drew , but maybe this could have a better name too?

My main concern here is legibility, took me a bit to figure out what b is in context of `createDefault`

File: cli/hack/test-e2e.sh (hunk 4 above)
Comment (praveenrewar) [reply cli/hack/test-e2e.sh:3]: Yep, that's correct.

**Comment by praveenrewar** [comment]
> Since this PR has a few refactors/improvements separated by commits I was wondering if it would be worth cleaning up some of the nomenclature in default_values.go

I have done the refactoring in the bash script only and tried to keep the changes in the other files to a minimal. I wanted to do more refactoring in `default-values.go` but I feel that a separate PR would make more sense?

**Comment by 100mik** [comment]
I think I am fine with a separate PR for this stuff
Track this somewhere maybe?

**Review by 100mik — APPROVED** [review]
LGTM!

**Comment by praveenrewar** [comment]
> Track this somewhere maybe?

I will be creating a new PR soon.
